internal/pwru/skb_metadata.go

```diff
@@ -0,0 +1,163 @@
+// SPDX-License-Identifier: Apache-2.0
+/* Copyright Leon Hwang */
+/* Copyright Authors of Cilium */
+
+package pwru
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/Asphaltt/mybtf"
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/asm"
+	"github.com/cilium/ebpf/btf"
+	"github.com/leonhwangprojects/bice"
+)
+
+const (
+	labelSetSkbMetadataExit = "__set_skb_metadata_exit"
+	setSkbMetadataStub      = "set_skb_metadata"
+	maxSkbMetadata          = 4
+)
+
+type SkbMetadata struct {
+	expr string
+	last string
+	t    btf.Type // type of last field
+	insn asm.Instructions
+}
+
+func isChar(c byte) bool {
+	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')
+}
+
+func isDigit(c byte) bool {
+	return c >= '0' && c <= '9'
+}
+
+func isValidChar(c byte) bool {
+	return isChar(c) || c == '_' || isDigit(c)
+}
+
+func newSkbMetadata(expr string, spec *btf.Spec) (*SkbMetadata, error) {
+	var md SkbMetadata
+	md.expr = expr
+
+	if !strings.HasPrefix(expr, "skb") {
+		return nil, errors.New("--output-skb-metadata must starts with skb")
+	}
+
+	for i := len(expr) - 1; i >= 0; i-- {
+		if !isValidChar(expr[i]) {
+			md.last = expr[i+1:]
+			break
+		}
+	}
+
+	types, err := spec.AnyTypesByName("sk_buff")
+	if err != nil {
+		return nil, err
+	}
+
+	skb := types[0]
+	ptr := &btf.Pointer{Target: skb}
+
+	res, err := bice.Access(bice.AccessOptions{
+		Expr:      expr,
+		Type:      ptr,
+		Src:       asm.R1,
+		Dst:       asm.R0,
+		LabelExit: labelSetSkbMetadataExit,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse expr %s: %w", expr, err)
+	}
+
+	size, err := btf.Sizeof(res.LastField)
+	if err != nil {
+		return nil, err
+	}
+	if size > 8 {
+		return nil, fmt.Errorf("skb metadata field '%s' (type %v) is too large, max 8 bytes", md.last, res.LastField)
+	}
+
+	md.t = res.LastField
+	md.insn = res.Insns
+	return &md, nil
+}
+
+func ParseSkbMetadataExprs(exprs []string, spec *btf.Spec) ([]*SkbMetadata, error) {
+	var md []*SkbMetadata
+
+	for _, e := range exprs {
+		if e == "" {
+			continue
+		}
+
+		m, err := newSkbMetadata(e, spec)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse skb metadata expr %s: %w", e, err)
+		}
+
+		md = append(md, m)
+	}
+
+	if len(md) > maxSkbMetadata {
+		return nil, fmt.Errorf("too many skb metadata exprs, max %d", maxSkbMetadata)
+	}
+
+	return md, nil
+}
+
+func InjectSetSkbMetadata(prog *ebpf.ProgramSpec, md []*SkbMetadata) error {
+	if len(md) == 0 {
+		return nil
+	}
+
+	var insns asm.Instructions
+	if len(md) > 1 {
+		insns = append(insns,
```

Comment:
Why is it not needed when len(md) = 1?

Reply:
if 'len(md) == 1', the **R1** already is `skb`.

As for 'md[1:]', it has to backup and recover **R1** with other register like **R7**.